runtime/runtime/src/genesis.rs

```diff
@@ -71,111 +68,211 @@ impl<'a> StorageComputer<'a> {
-            state_update.finalize().expect("Genesis state update failed");
-        let mut store_update = tries.store_update();
-        *current_state_root = tries.apply_all(&trie_changes, shard_uid, &mut store_update);
+        // Temporarily store a trie update with the "wrong" state root. This will be replaced
+        // shortly as soon as we know the new state root.
+        let old_state_update = std::mem::replace(
+            state_update,
+            self.tries.new_trie_update(self.shard_uid, *state_root),
+        );
+        let (_, trie_changes, state_changes) =
+            old_state_update.finalize().expect("Genesis state update failed");
+        let mut store_update = self.tries.store_update();
+        *state_root = self.tries.apply_all(&trie_changes, self.shard_uid, &mut store_update);
         if cfg!(feature = "protocol_feature_flat_state") {
             FlatStateChanges::from_state_changes(&state_changes)
-                .apply_to_flat_state(&mut store_update, shard_uid);
+                .apply_to_flat_state(&mut store_update, self.shard_uid);
         }
         drop(state_changes); // silence compiler when not protocol_feature_flat_state
         store_update.commit().expect("Store update failed on genesis initialization");
+        *state_update = self.tries.new_trie_update(self.shard_uid, *state_root);
+        *changes = 0;
+        *state_root
     }
+}
 
+impl<'a> Drop for AutoFlushingTrieUpdate<'a> {
+    fn drop(&mut self) {
+        self.flush();
+        self.active_writers.fetch_sub(1, atomic::Ordering::Relaxed);
+    }
+}
+
+pub struct GenesisStateApplier {}
+
+impl GenesisStateApplier {
     fn apply_batch(
-        current_state_root: &mut StateRoot,
+        storage: &mut AutoFlushingTrieUpdate,
         delayed_receipts_indices: &mut DelayedReceiptIndices,
-        tries: &mut ShardTries,
         shard_uid: ShardUId,
         validators: &[(AccountId, PublicKey, Balance)],
         config: &RuntimeConfig,
         genesis: &Genesis,
-        batch_account_ids: HashSet<&AccountId>,
+        account_ids: HashSet<AccountId>,
     ) {
-        let mut state_update = tries.new_trie_update(shard_uid, *current_state_root);
         let mut postponed_receipts: Vec<Receipt> = vec![];
-
         let mut storage_computer = StorageComputer::new(config);
-
+        tracing::info!(
+            target: "runtime",
+            ?shard_uid,
+            "processing records…"
+        );
         genesis.for_each_record(|record: &StateRecord| {
-            if !batch_account_ids.contains(state_record_to_account_id(record)) {
+            if !account_ids.contains(state_record_to_account_id(record)) {
                 return;
             }
-
             storage_computer.process_record(record);
-
-            match record.clone() {
-                StateRecord::Account { account_id, account } => {
-                    set_account(&mut state_update, account_id, &account);
-                }
+            match record {
```

Comment:
nit: I think we can move `storage.modify` outside the match for all cases except `postponed_receipts` and it would make logic slightly simpler.

Reply:
Not all of the branches modify the storage, though. Back when `modify` locked a mutex this current approach actually was the only sensible option. In that context handling a postponed or delayed receipt would result in an unnecessary locks being taken. Given that future modifications may still end up using a `Mutex` for one reason or another (account storage computation is pretty slow, for example) I think the minimal scoping of this approach is ideal.

Reply:
I see! Currently number of items corresponding to congested receipts is negligible, but who knows how big congestion will be in the future.